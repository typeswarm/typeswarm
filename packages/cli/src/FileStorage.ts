import { readFile, writeFile } from 'fs/promises';
import { injectable } from 'inversify';

export interface IFileStorage {
    read(fileName: string): Promise<string>;
    write(fileName: string, contents: string): Promise<void>;
}

@injectable()
export class FileStorageImpl_Real implements IFileStorage {
    read(fileName: string): Promise<string> {
        return readFile(fileName, 'utf8');
    }
    write(fileName: string, contents: string): Promise<void> {
        return writeFile(fileName, contents, 'utf8');
    }
}

@injectable()
export class FileStorageImpl_Mock implements IFileStorage {
    public files: Record<string, string> = {};

    read(fileName: string): Promise<string> {
        if (!this.files[fileName]) {
            throw new Error(`File not found: ${fileName}`);
        }
        return Promise.resolve(this.files[fileName]);
    }

    write(fileName: string, contents: string): Promise<void> {
        this.files[fileName] = contents;
        return Promise.resolve();
    }
}
