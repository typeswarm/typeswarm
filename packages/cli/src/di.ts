export const Types = {
    Logger: Symbol(),
    EntitiesProcessor: Symbol(),
    ComposeBuilder: Symbol(),
    CLIApplication: Symbol(),
};
