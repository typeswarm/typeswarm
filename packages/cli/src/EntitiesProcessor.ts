import 'reflect-metadata';

import {
    DefinitionsConfig,
    DefinitionsSecret,
    PropertiesConfigs,
} from '@typeswarm/core';
import {
    StrictServicesDict,
    StrictSpecification,
} from '@typeswarm/core/lib/normalize';
import { getHash } from '@typeswarm/core/lib/utils';

import { inject, injectable } from 'inversify';
import { basename, join, relative } from 'path';
import { Logger } from 'tslog';
import { Types } from './di';
import { IFileStorage } from './FileStorage';

@injectable()
export class EntitiesProcessor {
    constructor(
        @inject(Types.Logger)
        private logger: Logger,
        @inject(Types.IFileStorage)
        private fileStorage: IFileStorage
    ) {}

    private async processSingleEntity(
        name: string,
        entity: DefinitionsConfig | DefinitionsSecret,
        directory: string
    ): Promise<{ rotatedEntityName: string; rotatedEntityFile: string }> {
        if (!entity.file && !entity.data) {
            throw new Error(
                '.file or .data is missing in ' + JSON.stringify(entity)
            );
        }
        //TODO: support raw binary data
        const contents = entity.data
            ? entity.data
            : await this.fileStorage.read(entity.file as string);

        const hash = getHash(contents);
        const rotatedEntityFileBaseName = entity.data
            ? `${hash}.${name}.config` //TODO: slugify name
            : `${hash}.${basename(entity.file as string)}`;

        const rotatedEntityFile = join(directory, rotatedEntityFileBaseName);
        this.logger.info('Write file', rotatedEntityFile);
        //TODO: support raw binary data
        await this.fileStorage.write(rotatedEntityFile, contents);

        const rotatedEntityName = `${name}_${hash}`;
        return {
            rotatedEntityName,
            rotatedEntityFile,
        };
    }

    async processEntities(
        entity: 'secrets' | 'configs',
        spec: StrictSpecification,
        directory: string
    ): Promise<StrictSpecification> {
        const {
            //secrets or configs
            [entity]: entitiesMap = {},
            services = {},
        } = spec;
        const rotatedEntities: PropertiesConfigs = {};
        const rotatedServices: StrictServicesDict = {};

        const rotationMap: { [configName: string]: string } = {};

        for (const [name, config] of Object.entries(entitiesMap)) {
            if (
                config.name ||
                config.external ||
                !(config.file || config.data)
            ) {
                rotatedEntities[name] = config;
                continue;
            }
            const {
                rotatedEntityName: rotatedEntityName,
                rotatedEntityFile: rotatedEntityFile,
            } = await this.processSingleEntity(name, config, directory);
            rotatedEntities[rotatedEntityName] = {
                ...config,
                file: relative(directory, rotatedEntityFile),
            };
            delete rotatedEntities[rotatedEntityName].data;
            rotationMap[name] = rotatedEntityName;
        }

        for (const [name, service] of Object.entries(services)) {
            const {
                //service.configs or service.secrets
                [entity]: entityReferences = [],
            } = service;
            const rotatedEntityReferences: typeof entityReferences = [];
            for (const entityRef of entityReferences) {
                if (
                    typeof entityRef === 'object' &&
                    entityRef.source &&
                    rotationMap[entityRef.source]
                ) {
                    rotatedEntityReferences.push({
                        ...entityRef,
                        source: rotationMap[entityRef.source],
                    });
                } else {
                    rotatedEntityReferences.push(entityRef);
                }
            }
            rotatedServices[name] = {
                ...service,
                [entity]: rotatedEntityReferences,
            };
            if (rotatedEntityReferences.length === 0) {
                delete rotatedServices[name][entity];
            }
        }
        return {
            ...spec,
            [entity]: rotatedEntities,
            services: rotatedServices,
        };
    }
}
