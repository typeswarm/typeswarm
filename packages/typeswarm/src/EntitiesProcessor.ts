import fs from 'fs';
import { inject, injectable } from 'inversify';
import { basename, join, relative } from 'path';
import { Logger } from 'tslog';
import {
    DefinitionsConfig,
    DefinitionsSecret,
    PropertiesConfigs,
} from './compose-spec';
import { Types } from './di';
import { StrictServicesDict, StrictSpecification } from './normalize';
import { getHash } from './utils';

@injectable()
export class EntitiesProcessor {
    constructor(
        @inject(Types.Logger)
        private logger: Logger
    ) {}

    private async processSingleEntity(
        name: string,
        entity: DefinitionsConfig | DefinitionsSecret,
        directory: string
    ): Promise<{ rotatedEntityName: string; rotatedEntityFile: string }> {
        if (!entity.file && !entity.data) {
            throw new Error(
                '.file or .data is missing in ' + JSON.stringify(entity)
            );
        }
        //TODO: support raw binary data
        const contents = entity.data
            ? entity.data
            : await fs.promises.readFile(entity.file as string, 'utf8');

        const hash = getHash(contents);
        const rotatedEntityFileBaseName = entity.data
            ? `${hash}.${name}.config` //TODO: slugify name
            : `${hash}.${basename(entity.file as string)}`;

        const rotatedEntityFile = join(directory, rotatedEntityFileBaseName);
        this.logger.info('Write file', rotatedEntityFile);
        //TODO: support raw binary data
        await fs.promises.writeFile(rotatedEntityFile, contents, 'utf8');

        const rotatedEntityName = `${name}_${hash}`;
        return {
            rotatedEntityName,
            rotatedEntityFile,
        };
    }

    async processEntities(
        entity: 'secrets' | 'configs',
        spec: StrictSpecification,
        directory: string
    ): Promise<StrictSpecification> {
        const {
            //secrets or configs
            [entity]: entitiesMap = {},
            services = {},
        } = spec;
        const rotatedEntities: PropertiesConfigs = {};
        const rotatedServices: StrictServicesDict = {};

        const rotationMap: { [configName: string]: string } = {};

        for (const [name, config] of Object.entries(entitiesMap)) {
            if (
                config.name ||
                config.external ||
                !(config.file || config.data)
            ) {
                rotatedEntities[name] = config;
                continue;
            }
            const {
                rotatedEntityName: rotatedEntityName,
                rotatedEntityFile: rotatedEntityFile,
            } = await this.processSingleEntity(name, config, directory);
            rotatedEntities[rotatedEntityName] = {
                ...config,
                file: relative(directory, rotatedEntityFile),
            };
            delete rotatedEntities[rotatedEntityName].data;
            rotationMap[name] = rotatedEntityName;
        }

        for (const [name, service] of Object.entries(services)) {
            const {
                //service.configs or service.secrets
                [entity]: entityReferences = [],
            } = service;
            const rotatedEntityReferences: typeof entityReferences = [];
            for (const entityRef of entityReferences) {
                if (
                    typeof entityRef === 'object' &&
                    entityRef.source &&
                    rotationMap[entityRef.source]
                ) {
                    rotatedEntityReferences.push({
                        ...entityRef,
                        source: rotationMap[entityRef.source],
                    });
                } else {
                    rotatedEntityReferences.push(entityRef);
                }
            }
            rotatedServices[name] = {
                ...service,
                [entity]: rotatedEntityReferences,
            };
            if (rotatedEntityReferences.length === 0) {
                delete rotatedServices[name][entity];
            }
        }
        return {
            ...spec,
            [entity]: rotatedEntities,
            services: rotatedServices,
        };
    }
}
