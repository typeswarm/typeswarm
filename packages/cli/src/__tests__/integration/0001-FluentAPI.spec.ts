import { Container } from 'inversify';
import { Logger } from 'tslog';
import {
    Service,
    Image,
    Cluster,
    Volume,
    ServiceVolume,
    Config,
    Secret,
    ServiceConfig
} from '@typeswarm/core';
import { Types } from '../../di';
import { EntitiesProcessor } from '../../EntitiesProcessor';
import { FileStorageImpl_Mock, IFileStorage } from '../../FileStorage';

function createContainer() {
    const di = new Container();
    di.bind<EntitiesProcessor>(Types.EntitiesProcessor).to(EntitiesProcessor);
    di.bind<IFileStorage>(Types.IFileStorage)
        .to(FileStorageImpl_Mock)
        .inSingletonScope();
    di.bind<Logger>(Types.Logger).toConstantValue(
        new Logger({ displayFilePath: 'hidden' })
    );
    return di;
}

describe('Integration-0001 Fluent API', () => {
    it('should process fluent config', () => {
        const databaseVolume = Volume('mariadb-data');
        const cacheVolume = Volume('redis-data');

        const websiteConfig = Config('website-config').json({ hello: 'world' });
        const websiteSecret = Secret('website-secret').yaml({
            password: 'secret'
        });

        const srvCache = Service('cache')
            .image(Image('redis').tag('6.0'))
            .volume(ServiceVolume('/data').source(cacheVolume));

        const srvDb = Service('database')
            .image(Image('mariadb').tag('10'))
            .volume(ServiceVolume('/data').source(databaseVolume));

        const srvWebsite = Service('website')
            .image(Image('wordpress').tag('7.1'))
            .env('DATABASE_HOST', `${srvDb.data.name}`)
            .env('CACHE', `redis:${srvCache.data.name}`)
            .config(ServiceConfig('/etc/conf.json').source(websiteConfig))
            .secret(ServiceConfig('/etc/password.yaml').source(websiteSecret));

        const spec = Cluster('3.7')
            .service(srvCache)
            .service(srvWebsite)
            .service(srvDb)
            .volume(databaseVolume)
            .secret(websiteSecret)
            .config(websiteConfig);
        expect(spec.data).toMatchSnapshot();
    });
});
