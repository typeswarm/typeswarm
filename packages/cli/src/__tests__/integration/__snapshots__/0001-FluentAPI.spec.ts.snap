// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration-0001 Fluent API should process fluent config 1`] = `
Object {
  "services": Object {
    "cache": Object {
      "image": "redis:6.0",
      "volumes": Array [
        Object {
          "source": "redis-data",
          "target": "/data",
          "type": "volume",
        },
      ],
    },
    "database": Object {
      "image": "mariadb:10",
      "volumes": Array [
        Object {
          "source": "mariadb-data",
          "target": "/data",
          "type": "volume",
        },
      ],
    },
    "website": Object {
      "environment": Object {
        "CACHE": "redis:cache",
        "DATABASE_HOST": "database",
      },
      "image": "wordpress:7.1",
    },
  },
  "version": "3.7",
  "volumes": Object {
    "mariadb-data": null,
  },
}
`;
